theano/tensor/tests/test_basic.py

```diff
@@ -2744,6 +2745,130 @@ def check(m):
+            f_symb_tuple = fn(indices_symb, shape)
+            np.testing.assert_equal(ref, f_array_tuple())
+            np.testing.assert_equal(ref, f_symb_tuple())
+
+            # shape given as an array
+            shape_array = np.array(shape)
+            f_array_array = fn(indices, shape_array)
+            f_symb_array = fn(indices_symb, shape_array)
+            np.testing.assert_equal(ref, f_array_array())
+            np.testing.assert_equal(ref, f_symb_array())
+
+            # shape given as a theano variable
+            shape_symb = theano.shared(shape_array)
+            f_array_symb = fn(indices, shape_symb, len(shape))
+            f_symb_symb = fn(indices_symb, shape_symb, len(shape))
+            np.testing.assert_equal(ref, f_array_symb())
+            np.testing.assert_equal(ref, f_symb_symb())
+
+            # shape testing
+            self._compile_and_check([],
+                                    unravel_index(indices, shape_symb, order=order, ndim=len(shape)),
+                                    [], UnravelIndex)
+
+        for order in ('C', 'F'):
+            for index_ndim in (0, 1, 2):
+                check((3,), index_ndim, order)
+                check((3, 4), index_ndim, order)
+                check((3, 4, 5), index_ndim, order)
+
+        # must specify ndim if length of dims is not fixed
+        self.assertRaises(TypeError, unravel_index, ivector(), ivector())
+
+        # must provide integers
+        self.assertRaises(TypeError, unravel_index, fvector(), (3, 4))
+        self.assertRaises(TypeError, unravel_index, (3, 4), (3.4, 3.2))
+        self.assertRaises(TypeError, unravel_index, (3, 4), (3, 3), ndim=5.4)
+
+        # dims must be a 1D sequence
+        self.assertRaises(TypeError, unravel_index, (3, 4), 3)
+        self.assertRaises(TypeError, unravel_index, (3, 4), ((3, 4),))
+
+
+class test_ravel_multi_index(utt.InferShapeTester):
+    def test_ravel_multi_index(self):
+        def check(shape, index_ndim, mode, order):
+            multi_index = np.unravel_index(np.arange(np.product(shape)), shape, order=order)
+            # create some invalid indices to test the mode
+            if mode in ('wrap', 'clip'):
+                multi_index = (multi_index[0] - 1,) + multi_index[1:]
+            # test with scalars and higher-dimensional indices
+            if index_ndim == 0:
+                multi_index = tuple(i[-1] for i in multi_index)
+            elif index_ndim == 2:
+                multi_index = tuple(i[:, np.newaxis] for i in multi_index)
+            multi_index_symb = [theano.shared(i) for i in multi_index]
+
+            # reference result
+            ref = np.ravel_multi_index(multi_index, shape, mode, order)
+
+            def fn(mi, s):
+                return function([], ravel_multi_index(mi, s, mode, order))
+
+            # shape given as a tuple
+            f_array_tuple = fn(multi_index, shape)
+            f_symb_tuple = fn(multi_index_symb, shape)
```

Comment:
I don't think it is very useful to compile with constant and shared variable. What about doing only one of them here and the other around line 2841? This would help speed up our tests that is already too long.

idem for the other test classe.